routing/localchans/manager.go

```diff
@@ -106,6 +125,40 @@ func (r *Manager) UpdatePolicy(newSchema routing.ChannelPolicy,
 		return err
 	}
 
+	// Return an error if any of the channels in chansToUpdate
+	// were invalid.
+	var errStr string
+	for chanPoint := range chansToUpdate {
+		channel, err := r.FetchChannel(chanPoint)
+		if errors.Is(err, channeldb.ErrChannelNotFound) {
+			errStr += fmt.Sprintf(
+				"channel %s:%d not found; ",
+				chanPoint.Hash.String(), chanPoint.Index,
+			)
+			continue
+		}
+		if err != nil {
+			errStr += fmt.Sprintf(
+				"error fetching channel %s:%d; ",
+				chanPoint.Hash.String(), chanPoint.Index,
+			)
+			continue
+		}
+		if channel.IsPending {
+			errStr += fmt.Sprintf("channel %s:%d is "+
+				"not yet confirmed; ",
+				chanPoint.Hash.String(), chanPoint.Index)
+			continue
+		}
+		errStr += fmt.Sprintf("channel %s:%d could "+
+			"not update its policies; ",
+			chanPoint.Hash.String(), chanPoint.Index)
+	}
+	if errStr != "" {
+		// Report error msg for both invalid and inactive channels.
+		return fmt.Errorf(errStr + errInactiveChans)
+	}
+
```

Comment:
Hmmm so the problem here is that if there were also some successful channel updates (`err := r.updateEdge(info.ChannelPoint, edge, newSchema)` above) then we still return here and then dont call `r.UpdateForwardingPolicies(policiesToUpdate)` below. (I think a test for this would be a good idea).

An idea would be to _first_ loop through all the channels-to-update and determine if there are any errors (wrong txid, inactive, pending) and just return an error and include in the error message that no updates were made to _any_ channels . Only if there are no bad apple channels then we go on to `err := r.ForAllOutgoingChannels(func(...`. So then it is an all-or-nothing situation (less ambiguity). Either that or keep track of all the problem channels, let the others update and then return an error naming the problem channels plus a message saying that all other channels were updated successfully. I lean toward the second option since if the user just wants to update all their channels but then a few are inactive then we dont want them to go type out the whole list of channels manually.

Reply:
Thanks for the review! I'll remember to run the tests before my next commit + fix the test model.
I agree, the second option is the way to go. I'll edit the error message to clarify what is happening in the background.